### PR TITLE
Fix for both issues 274 and 278

### PR DESCRIPTION
**Pull request by schweitzpgi**
Makes each subprogram/procedure a distinct scope and fixes the bug in stack initialization.

### CODE REVIEW

File: lib/semantics/resolve-labels.cc [hunk 1]
```diff
@@ -495,8 +494,9 @@ class ParseTreeAnalyzer {
     currentScope_ = programUnits_.back().scopeModel.size() - 1;
     return true;
   }
-  bool PushNewScope() {
+  bool InitializeNewScopeContext() {
```
Comment (klausler) [inline lib/semantics/resolve-labels.cc]: These pushed subscopes still get popped, right?

**Comment by schweitzpgi** [comment]
> LGTM if all tests still pass and bugs stayed fixed. Thanks.

I changed the algorithm a bit. Rather than popping back to the root at the END statement, the scope tree is left as created.  The next scope (if any) just forcibly resets the (pristine) tree state.

**Comment by klausler** [comment]
Is the scope tree still needed for anything?

**Comment by schweitzpgi** [comment]
> Is the scope tree still needed for anything?

Oh, yes. Very much so. All subsequent constraint checks may use it.

**Pull request merged**